Run Xcode 15 in CI and skip or disable tests that fail

mokagio opened the pull request:
A bit of an underwhelming PR, where instead of fixing things, I work around failures by disabling tests 😞 In my defense, there doesn't seem much we can do about how the existing implementation of the problematic tests behaves with the iOS 17 SDK.

I thought I'd put this up as a starting point. [The unit tests passed in the draft PR](https://buildkite.com/automattic/wordpress-ios/builds/18506#_). Merging would at least let us officially use Xcode 15.0.1 as development toolchain, so that we can move on to fixing things and find new issues to tackle.

Open to suggestions on how to update the failing code.

## Regression Notes

1. Potential unintended areas of impact – N.A.
2. What I did to test those areas of impact (or what existing automated tests I relied on) – N.A.
3. What automated tests I added (or what prevented me from doing so) – N.A.

---

**PR submission checklist:**

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. **N.A.**
- [x] I have considered adding accessibility improvements for my changes. **N.A.**
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. **N.A.**

**UI changes testing checklist:** Not a UI PR.

mokagio committed:
Update `TimeZoneFormatterTests` to pass on Xcode 15

mokagio committed:
Update `TimeZoneFormatterTests` to pass on Xcode 15

mokagio committed:
Use Xcode 15.0.1 in CI

mokagio committed:
Use Xcode 15.0.1 in CI

mokagio committed:
Remove a useless reassignment in `WKCookieJarTests`

mokagio committed:
Remove a useless reassignment in `WKCookieJarTests`

mokagio committed:
Mark `WKCookieJarTests` as expected to fail

mokagio committed:
Mark `WKCookieJarTests` as expected to fail

mokagio committed:
Remove redundant custom test assertion, which was also out of sync

mokagio committed:
Remove redundant custom test assertion, which was also out of sync

mokagio committed:
Skip two tests failing because of a changed `WKWebView` behavior

mokagio committed:
Skip two tests failing because of a changed `WKWebView` behavior

kean commented:
I found the following issue when testing on Xcode 15 with the new iOS SDK: https://github.com/wordpress-mobile/WordPress-iOS/issues/21649 (blocker). There may be more issues of a similar kind. I'll require a bit more effort to migrate to Xcode 15 because it also means migrating to iOS 17, which the app currently doesn't (fully) support.

mokagio commented:
Thanks @kean 👍 

I subscribed to the issue. I also created an [Xcode 15 and iOS 17 project in this repo](https://github.com/orgs/wordpress-mobile/projects/212/) as a quick way to collect and monitor progress on this kind of issues. Feel free to add to it, thanks!

mokagio committed:
Update Fastlane to version 2.217.0 to get Xcode 15 trainer fix

See https://github.com/fastlane/fastlane/pull/21493

mokagio committed:
Update Fastlane to version 2.217.0 to get Xcode 15 trainer fix

See https://github.com/fastlane/fastlane/pull/21493

mokagio commented on WordPress/WordPressTest/WKCookieJarTests.swift:
```diff
@@ -3,7 +3,7 @@ import WebKit
 
 class WKCookieJarTests: XCTestCase {
-    var wkCookieStore = WKWebsiteDataStore.nonPersistent().httpCookieStore
+    var wkCookieStore: WKHTTPCookieStore!
```
The value is being assigned in `setUp`. It's best to do it like that anyway, so we get a fresh one for every test.

pachlava committed:
[UI Tests] Replaced "4 AM" with "4 AM".

pachlava committed:
[UI Tests] Replaced "4 AM" with "4 AM".

kean commented:
LGTM.

The fix for post scheduling in iOS 17 was also fixed https://github.com/wordpress-mobile/WordPress-iOS/pull/22012. I've been running Xcode 15 for a couple of weeks now with no issues. I also need Xcode 15 for the following PR https://github.com/wordpress-mobile/WordPress-iOS/pull/22060, so I would appreciate if we could get this into `trunk` sooner rather than later.

mokagio commented:
Thanks @pachlava for having a look at the UI tests! 🙌 

I'll share here a summary of an internal conversation for visibility: Sergiy noticed that the UI tests are super slow and, because of the slowness, more flaky and prone to fail. (_Talk about selective attention, I looked at every single CI build but was only focused on the unit tests and was totally oblivious to how long the UI tests took!_) The UI tests alone can now take more than 2 hours to run, that's too long.

Unfortunately, this is [a known performance issue with Xcode 15 in CI](https://discuss.circleci.com/t/severe-performance-problems-with-xcode-15/49205/125) for which there doesn't seem to be a fix yet.

Given the performance implications, I see only two options:

1. Disable the UI tests, which we don't want to do because we care about what they tell us; or
2. Delay adopting Xcode 15, which we don't want to do because of the need to address iOS 17 issues

As much as it pains me, I think we might need to disable the UI tests in CI. Nothing should stand in the way of fixing user issues.

But before doing that, it is also possible that our new M1 CI infra will help with this, based on some of the discussion in the link above and on [this forums thread](https://developer.apple.com/forums/thread/735177?page=2) Apple Silicon hardware fairs better. That's not surprising, of course, in terms of raw performance, but there might be something structural going on there, too. cc @jkmassel

kean commented:
I'll add one more potential option for consideration: run UI tests once a day and deliver a report to the team in case they fail. This will alleviate 50%+ of the pressure from the CI. If it's currently feasible to run UI tests for every build, it means there are not enough of them.

jkmassel commented:
I'd like to see how this performs on Apple Silicon – it's great that this PR targets `15.0.1`, because it means that we're immediately compatible and can easily switch back and forth.

I'd be surprised if the performance issues hit us too badly there, from the thread it seems like it's mostly folks who are using Rosetta to run their tests, but I didn't read in-depth. If the tests continue to take an unacceptably long amount of time we can subdivide them further – UI tests are pretty parallelizable. 

If none of that works, I'd really like to dig into the tests and find out exactly what they're doing, and if this behaviour can be replicated locally – maybe it's something to do with running under a VM?  We've seen enormous instability with Xcode 15.0 for Tumblr, so the idea that performance on Intel is somehow compromised in that version isn't too surprising. The good news is that Apple Silicon made an enormous difference there, so hopefully it will here too?

I'd very much like to avoid turning off a test suite unless we're left with no other option – and running it on some kind of schedule is essentially turning it off. Every time we've tried it we've ended up in a situation where the tests are broken the majority of the time, so even legitimate failures are dismissed as flakiness. If we can't run a test on every PR, that test isn't worth having. And if we need more capacity to run those tests, we'll add it – computers are cheap, people's time is expensive, so we run the tests – as often as we can.

pachlava commented:
> run UI tests once a day and deliver a report to the team in case they fail.

@kean 👋 The issue with UI tests after upgrade to Xcode 15 is not the slowness alone, it's the fact that slowness causes timeouts, resulting in them always failing. One alternative is running them locally against `trunk` daily, which is probably better than nothing.

kean commented:
The current UI test suite has only 35 tests, which is not a lot of tests tbh, and it already overloads the CI. From the developer side, if it takes more than 10 minutes, I'd say it shouldn't run on every push or it will hinder the development flow and works against short-lived branches and TBD. 

> The good news is that Apple Silicon made an enormous difference there, so hopefully it will hear too?

How long does it take to run the tests on M2/M3 locally? I've yet to run the tests locally because you need a Java runtime. I begrudgingly installed it, ran `rake mocks`, turned off the proxy, but I am still yet to be able to run them 😞 

> daily, which is probably better than nothing.

That's how it's traditionally been done, otherwise, it sets the wrong incentives. The incentive should be to allow the automation team to create as many UI tests as possible and not worry about their performance as long as it takes no more than 8-12 hours to run the entire suite.

pachlava commented:
>  If the tests continue to take an unacceptably long amount of time we can subdivide them further – UI tests are pretty parallelizable.

@jkmassel, parallelization is already in use for these UI tests (if we're talking about the same kind of parallelization), and it's one of the reasons for fails. One of the frequent issues on CI (with Xcode 15) is inability to launch additional Simulators:

<img width="914" alt="Screenshot 2023-11-20 at 21 41 06" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/73365754/602c9191-1ad1-4be7-9120-d21cc0f22a1e">

I'll actually check if turning off parallelization (just for test) improves the stability on CI.

> If none of that works, I'd really like to dig into the tests and find out exactly what they're doing, and if this behaviour can be replicated locally – maybe it's something to do with running under a VM?

From my testing yesterday, they always passed locally (I was running the same Xcode 15.0.1), and always failed on CI. This goes in line with things described in the [CircleCI thread](https://discuss.circleci.com/t/severe-performance-problems-with-xcode-15/49205/35), with tests working well on CI with Xcode 14.3 and Xcode 15 _locally_, but failing for Xcode 15 when on CI.

pachlava commented:
@kean 👋 🙂 

> The current UI test suite has only 35 tests, which is not a lot of tests tbh, and it already overloads the CI. From the developer side, if it takes more than 10 minutes, I'd say it shouldn't run on every push. It hinders the development flow and works against short-lived branches and TBD.

The question of the tests speed/quantity balance has been the focus of @wordpress-mobile/mobile-ui-testing-squad work for long time, and even though the value of 30 minutes has been frequently mentioned by Engineers as a psychological limit, we try to keep them as fast as possible by working on the framework and involving WireMock and parallelization. The current usual duration of UI tests on Buildkite is 20 minutes.

> How long does it take to run the tests on M2/M3 locally?

The local execution on Intel is 12 minutes, and for Silicon numbers I'd ask @jostnes or @tiagomar to chime in, since I'm yet to update my laptop.

> That's how it's traditionally been done, otherwise, it sets the wrong incentives. The incentive should be to allow the automation team to create as many UI tests as possible and not worry about their performance as long as it takes no more than 8-12 hours to run the entire suite.

I know where you're coming from, and this has been my previous experience as well, in the companies where the responsibility for the quality was either mainly or entirely on the testing teams (aka being a fishing net, or the last line of defence). This question is approached differently here, because of several reasons:

- Having hundreds/thousands of UI tests does not scale well. With the current amount of people in the QE team, and the number of apps we work on (7 apps, each runs on two platforms), we focus on covering the critical flows only. An example of this for WPiOS is [here](https://github.com/wordpress-mobile/WordPress-iOS/tree/trunk/WordPress/UITests#test-coverage).
- More importantly, it's hard to measure a positive impact (thus, justify spending time) of having hundreds of UI tests VS having only the critical paths automated. In both cases, the impact of UI test is evident only when they find a real issue.
- Running the tests on every push  / every PR makes it easier to trace back a potential issue as opposed to having them executed on schedule. The stabilization efforts of @wordpress-mobile/mobile-ui-testing-squad allowed making the passing UI tests mandatory for merge to `trunk`, and it helped us avoid a number of crashes or serious issues being merged.
- With actively developed apps, the introduced changes often require UI tests to be updated to match the current state of the app. The shortest path for doing this is noticing the fact of test breakage right in the PR, and fixing it right there. This is often done by Engineering teams, in some cases this is done by QE team.
- Last but not the least, the initiative of moving the quality processes to the earlier stages of development process (e.g. with "QE in Dev" - PdcxQM-Gy-p2) shifts the mentality from "testing all the things in the end" (no matter if we mean PR end, day end or sprint end) to "making sure the bugs don't get implemented at all". Both cases are theoretical extremes, but they illustrate the reasoning behind keeping the QE teams small and number of UI tests relatively low.

pachlava committed:
[Temp Experiment] See if turning off parallelization helps with UI tests issues.

pachlava committed:
[Temp Experiment] See if turning off parallelization helps with UI tests issues.

kean commented:
I appreciate the detailed answer, @pachlava. It's a totally understandable position, considering the constraint in resources. If expanding the suite is a no-goal, and it doesn't take a lot of time to run, it makes total sense to run them on every branch. I mainly just commented to offer a potential alternative temporary solution – I'm interested in getting Xcode 15 support merged before the 23.8 soft freeze and the holidays (Nov 23 in the US).

tiagomar commented:
>  The local execution on Intel is 12 minutes, and for Silicon numbers I'd ask @jostnes or @tiagomar to chime in, since I'm yet to update my laptop.

I'm still running on Intel 😬 . I know Jos has recently gotten a new machine, but she's afk for most of this week.

pachlava committed:
Revert "[Temp Experiment] See if turning off parallelization helps with UI tests issues."

This reverts commit 53ea3277cb7a8a0def7a7c6d4a46e05c523cfe86.

pachlava committed:
Revert "[Temp Experiment] See if turning off parallelization helps with UI tests issues."

This reverts commit 53ea3277cb7a8a0def7a7c6d4a46e05c523cfe86.

pachlava commented:
>  I'm interested in getting Xcode 15 support merged before the 23.8 soft freeze and the holidays

@kean, I'm all for this as well, even knowing that we'll need to have UI tests disabled for the time being.

jkmassel commented:
@pachlava – just to clarify, in this case I'm referring to parallelizing by running on multiple machines at once, not by running multiple simulators on the same machine

Hope that helps!

crazytonyli commented on WordPress/WordPressTest/ViewRelated/Tools/Time Zone/TimeZoneFormatterTests.swift:
```diff
@@ -38,13 +38,30 @@ class TimeZoneFormatterTests: XCTestCase {
+            XCTAssertEqual("6:00\u{202F}PM", timeAtZone)
+        } else {
+            XCTAssertEqual("6:00 PM", timeAtZone)
+        }
```
You should be able to unify the two assertions by using regex match: `6:00\s+PM`? The unit test should not care about the exact bytes of the whitespace, as long as it's a whitespace, right?

crazytonyli commented on WordPress/WordPressTest/ViewRelated/Tools/Time Zone/TimeZoneFormatterTests.swift:
```diff
@@ -38,13 +38,30 @@ class TimeZoneFormatterTests: XCTestCase {
+            XCTAssertEqual("7:00\u{202F}PM", timeAtZone)
+        } else {
+            XCTAssertEqual("7:00 PM", timeAtZone)
+        }
```
Same as the above.

mokagio committed:
Remove git-cache plugin to run on Apple Silicon CI

mokagio committed:
Run UI tests on "flagship" Simulators for iOS 17

See if this makes any difference on the execution time in CI

mokagio commented:
Good news / Bad news: Preliminary experiments show that in the Apple Silicon CI, the UI tests run and pass with more or less the usual unfortunate flakiness. Bad news, they are still slow, running in the order of 30-40minutes.
Example https://buildkite.com/automattic/wpios-macv2-test/builds/11

Stay tuned

mokagio committed:
Run UI tests on iPhone 15

mokagio commented:
Unrelated to CI speed etc, but I think the failure in `test_fetchPrompts_passesTheDateCorrectly` that has appeared in 2 out of 3 runs so far is due to inconsistent expectations on the date formatter used between tests and production.

In the screenshot, you can see the tests create a date from "2022-01-02" in UTC which results to 2022-01-01 13:00:00. But the production code uses the local time zone and the date string it computes is "2022-01-02"

<img width="1978" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1218433/53f2fd25-a1e5-48f0-9335-fd60eb26c10b">

mokagio committed:
Revert "Revert "[Temp Experiment] See if turning off parallelization helps with UI tests issues.""

This reverts commit 244da46553ebf6274c877ddd3e7885ce1f1bc1b2.

mokagio committed:
Use UTC date formatter in `BloggingPromptsServiceTests`

This is because some of the expectations use UTC to compute the date,
resulting in edge cases based on the local time zone where the day
locally is different from the UTC one.

mokagio committed:
Always track screen captures in UI tests

mokagio commented:
@pachlava and everyone else who might have checked out this branch locally: Apologies for the force push which will require a force pull on your end, I didn't realize I was on this branch that had already been shared when I run the command. 😞

mokagio committed:
Disable parallel testing in Fastlane, too

mokagio committed:
Set unit tests to run in English and with region US

This is a followup to 5ee54f6ffb987de62b79da3de4cf547d11aabbca, which
fixed the issue on my end but not in CI.

mokagio commented:
Good news. With parallelism disabled in Fastlane as well, b9ee1679a1fee27f35c, the UI tests seems stable and not as slow in the Apple Silicon CI. I think the performance gain we get on builds and unit tests more than makes up for the – hopefully temporary – slowdown on UI tests. 

See https://buildkite.com/automattic/wpios-macv2-test/builds/15 and https://buildkite.com/automattic/wpios-macv2-test/builds/16

If we are in a rush to deploy the Apple Silicon CI on an app and/or Xcode 15.0.1 support, we could merge this PR as is. Otherwise, I'll try to tidy it up tomorrow as well as further investigating some failures. 

cc @jkmassel @pachlava

pachlava committed:
Merge branch 'mokagio/fix-code-15-tests' of https://github.com/wordpress-mobile/WordPress-iOS into mokagio/fix-code-15-tests

pachlava committed:
Wait for comment text to load for longer.

pachlava committed:
[UI Tests] On CI it takes even longer than 10s to load the comment.

pachlava commented:
Just leaving a note that starting with 4b38795c8c1fc1c45855464f233c3ee8b5fd7d25 the UI tests started failing again (not related to the commit I believe). The situation with iPhone is much better: only `testLikeNotification` fails, and I hope that 0f32cb9f93e4f00f9cef9706f42f4f17471a0d10 will help with that. On iPad, it's much sadder, and generally the reason is the same as for the fail in said `testLikeNotification`: extremely slow loading of screens or their parts. For example, it takes 30 seconds to load the empty editor:

https://github.com/wordpress-mobile/WordPress-iOS/assets/73365754/1cf9e9e9-9a83-48c6-a531-ae8c4621d0e7

And this translates to other screens as well. If we manage to have iPhone tests running, I think disabling iPad ones is a better alternative than increasing all timeouts x2 or x3, until this is fixed by Apple (am I too naive to think that this is going to happen 🤔).

jostnes commented:
> How long does it take to run the tests on M2/M3 locally?

I'm on M2 Max, running the test suite locally, when parallel execution is turned off, the test suite takes ~18 minutes. With Xcode's parallel execution turned on, using two simulators, it was slightly under 10 minutes.

For both cases, though there were failures when running locally, which was handled by the retries (set to 3), the time taken to complete could be lower if there were no failures. So it seems like the flakiness is not only on CI but is also present on local runs when using this branch:
![image](https://github.com/wordpress-mobile/WordPress-iOS/assets/17252150/cc742492-a0cc-49c9-b2f6-d437238b9518)

(above test on iPhone 15, Xcode 15.0)

pachlava committed:
[UI Tests] Changed an element locator.

pachlava commented:
After some playing around, I could fix the `testLikeNotification` for CI, so the iPhone job passes now, but the iPad one still fails (more https://github.com/wordpress-mobile/WordPress-iOS/pull/21921#issuecomment-1824902941)

@jostnes, I had to change the comment text locator in 655a424c34de8a32203bd1f97a261d0e893f51da, because for whatever reason, on CI the comment text is not discoverable at all now. Locally (and before on CI) it's a child of a WebView, now it's just not locatable on CI (objects snapshot from CI):

https://github.com/wordpress-mobile/WordPress-iOS/assets/73365754/51467148-1a36-44ee-a50d-f55903e89db4

jostnes commented:
> @jostnes, I had to change the comment text locator in https://github.com/wordpress-mobile/WordPress-iOS/commit/655a424c34de8a32203bd1f97a261d0e893f51da, because for whatever reason, on CI the comment text is not discoverable at all now.

Thank you for making the change @pachlava! Also I agree with you that we could probably stop running the tests on iPad since it's extremely slow and tests keep failing because of time-outs, and keep it iPhone only for the time being. But I also wonder if this slowness is only seen on CI or will it be seen by actual users on a slower network as well after this change?

pachlava commented:
@jostnes, from my understanding of the issue, this is related to virtualization that is used on CI (affects both BK and CircleCI). It's not related to the network speed.

mokagio committed:
Merge 'origin/trunk' into mokagio/fix-code-15-tests

mokagio committed:
Use Xcode 15.1

mokagio committed:
Fix indentation in `TestObserver`

jostnes committed:
Merge branch 'trunk' into mokagio/fix-code-15-tests

mokagio committed:
Add workaround for iOS 17.2 Simulator breaking passwords settings

jostnes committed:
temp remove TestObserver

jostnes committed:
temp remove auto retries

jostnes committed:
added workarounds around login

jostnes committed:
Merge branch 'mokagio/fix-code-15-tests' of https://github.com/wordpress-mobile/WordPress-iOS into mokagio/fix-code-15-tests

mokagio commented on WordPress/UITestsFoundation/Screens/Login/Unified/PasswordScreen.swift:
```diff
@@ -29,7 +29,7 @@ public class PasswordScreen: ScreenObject {
     public func proceedWithValidPassword() throws -> LoginEpilogueScreen {
         try tryProceed(password: "pw")
-
+        app.dismissSavePasswordPrompt()
```
@jostnes  I'm surprised by this. `tryProceed(password:)` already calls it. 

Did it make a difference or is it an experiment?

jostnes replied:
oh yeah, this has been removed, i overlooked this earlier, this was also the reason the test took a longer time to complete (as it was doing it twice, the second time completely unnecessary 😅)

mokagio commented on WordPress/UITests/JetpackUITests-Info.plist:
```diff
@@ -20,7 +20,5 @@
 	<key>CFBundleVersion</key>
 	<string>1</string>
-	<key>NSPrincipalClass</key>
-	<string>${PRODUCT_NAME}.TestObserver</string>
```
@jostnes I'd recommend keeping this setup and skipping the `executeWithRetries(disableAutoFillPasswords)` in `testBundleWillStart` from `TestObserver`. 

I'm saying this because it's easier to find the `TestObserver` setup in the code than in this `Info.plist` and hopefully we'll be able to go back to that setup when Apple fixes their Simulator bug (assuming my theory about it being a simulator bug is correct)

jostnes replied:
yup, this is a temporary removal to save testing time and skip the part where the test tries to turn it off and fails every time because of the simulator bug. this will be re-added before merging.

jostnes committed:
decrease timeout

jostnes committed:
remove unnecessary checks

jostnes committed:
fix logic for tapUntilCondition

jostnes committed:
turn retry back on

jostnes committed:
fix stats test by making year dynamic

jostnes committed:
add retry to dismiss save password prompt

mokagio commented:
Closing in favor of @jostnes' https://github.com/wordpress-mobile/WordPress-iOS/pull/22270
